### PR TITLE
Unregistered users can leave comments; registered users can change their rating

### PR DESCRIPTION
**Pull request by YukiAmeka**

### CODE REVIEW

File: modules/comment/comment.model.js [hunk 1]
```diff
@@ -3,16 +3,13 @@ const mongoose = require('mongoose');
...
   },
   date: {
     type: Date,
     default: Date.now,
   },
-  user: {
-    type: mongoose.Schema.Types.ObjectId,
-    ref: 'User',
-  },
+  email: String,
```
Comment (Alan18081) [inline modules/comment/comment.model.js:12]: Hey, I think, we have decided to have `user` object and nested there `{ email: String, name: String: adminId: ObjectId }`. Please, refactor it in the GraphQL model too

File: modules/rate/rate.model.js [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+  },
+  user: {
+    type: mongoose.Schema.Types.ObjectId,
+    ref: 'User',
+  },
+  product: {
+    type: mongoose.Schema.Types.ObjectId,
+    ref: 'Product',
+  },
+  show: Boolean,
```
Comment (Alan18081) [inline modules/rate/rate.model.js:20]: Why do we need this property?

File: modules/rate/rate.model.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
+const mongoose = require('mongoose');
+
+const rateSchema = new mongoose.Schema({
```
Comment (Alan18081) [inline modules/rate/rate.model.js:3]: I think it is not the best approach for the handling of ratings. From my perspective, it is better to have a field
`ratings: {   [userId]: [ratingValue]  }`. Per each product, we are going to have one rating map. Also, it is good to think, do we really need a separate collection for the ratings. Maybe it would be much easier to just include it inside the product model. Let's discuss it on the meeting

File: modules/comment/comment.model.js (hunk 1 above)
Comment (YukiAmeka) [reply modules/comment/comment.model.js:12]: I must have been absent when this was decided. Having said that, including name is a good idea, I'll do that. `adminId` on the other hand I don't quite understand - what is it supposed to be referring to and for what purpose?

File: modules/rate/rate.model.js (hunk 2 above)
Comment (YukiAmeka) [reply modules/rate/rate.model.js:20]: To be able to hide unfavorable ratings of course :)

File: modules/rate/rate.model.js (hunk 3 above)
Comment (YukiAmeka) [reply modules/rate/rate.model.js:3]: Ooh, this is a really good idea. Granted, it'll have to be `ratings: { user: [userId], value: [ratingValue] }`, as a changeable value of `userId` cannot serve as a key in schema, but incorporating it in products this way should be way easier to handle. I'll check if there seem to be no hidden issues and modify accordingly or get back to you. Thank you for the suggestion

File: modules/comment/comment.graphql.js [hunk 4]
```diff
@@ -3,7 +3,7 @@ type Comment {
 _id: ID!
 text: String!
 date: String!
-user: User!
+user: UserForComment
```
Comment (VolodymyrTrach) [inline modules/comment/comment.graphql.js:6]: add "!" to make this field required

File: modules/comment/comment.graphql.js [hunk 5]
```diff
@@ -12,6 +12,7 @@ show: Boolean
 const commentInput = `
 input commentInput {
   text: String
+  user: UserForCommentInput
```
Comment (VolodymyrTrach) [inline modules/comment/comment.graphql.js:15]: add "!" to make this field required

File: modules/product/product.model.js [hunk 6]
```diff
@@ -55,12 +55,20 @@ const productSchema = new mongoose.Schema({
...
-    {
+  userRates: [{
+    user: {
       type: mongoose.Schema.Types.ObjectId,
       ref: 'User',
     },
-  ],
+    rate: {
+      type: Number,
+      default: 0,
```
Comment (VolodymyrTrach) [inline modules/product/product.model.js:65]: take out the default value, the user should decide it.

File: modules/user/user.graphql.js [hunk 7]
```diff
@@ -26,7 +26,7 @@ input UserInput {
     lastName: String
     password: String
     role: String
-    email: String
+    email: String!
```
Comment (VolodymyrTrach) [inline modules/user/user.graphql.js:29]: undo this change, if the user will log in form facebook - he should be able to register

File: typeDefs.js [hunk 8]
```diff
@@ -127,6 +127,19 @@ const typeDefs = gql`
     additionalPrice: Int
   }
 
+  type UserForComment {
+    email: String
```
Comment (VolodymyrTrach) [inline typeDefs.js:131]: add "!" to make this field required

File: typeDefs.js [hunk 9]
```diff
@@ -127,6 +127,19 @@ const typeDefs = gql`
...
+    name: String
+    images: ImageSet
+    isAdmin: Boolean
+  }
+
+  type UserRate {
+    user: User
```
Comment (VolodymyrTrach) [inline typeDefs.js:138]: add "!" to make this field required

File: typeDefs.js [hunk 10]
```diff
@@ -227,6 +240,17 @@ const typeDefs = gql`
     buildingNumber: String
     appartment: String
   }
+  input UserForCommentInput {
+    email: String
```
Comment (VolodymyrTrach) [inline typeDefs.js:244]: add "!" to make this field required

File: typeDefs.js [hunk 11]
```diff
@@ -227,6 +240,17 @@ const typeDefs = gql`
...
+    name: String
+    images: ImageSetInput
+    isAdmin: Boolean
+  }
+  input UserRateInput {
+    user: UserInput
```
Comment (VolodymyrTrach) [inline typeDefs.js:250]: I think it should be an ID!

File: modules/comment/comment.graphql.js (hunk 4 above)
Comment (YukiAmeka) [reply modules/comment/comment.graphql.js:6]: Done

File: modules/comment/comment.graphql.js (hunk 5 above)
Comment (YukiAmeka) [reply modules/comment/comment.graphql.js:15]: Done

File: modules/product/product.model.js (hunk 6 above)
Comment (YukiAmeka) [reply modules/product/product.model.js:65]: Done

File: typeDefs.js (hunk 8 above)
Comment (YukiAmeka) [reply typeDefs.js:131]: Done

File: typeDefs.js (hunk 9 above)
Comment (YukiAmeka) [reply typeDefs.js:138]: Done

File: typeDefs.js (hunk 10 above)
Comment (YukiAmeka) [reply typeDefs.js:244]: Done

File: modules/user/user.graphql.js (hunk 7 above)
Comment (YukiAmeka) [reply modules/user/user.graphql.js:29]: We discussed it yesterday and discarded an option of no email, didn't we?..

File: typeDefs.js (hunk 11 above)
Comment (YukiAmeka) [reply typeDefs.js:250]: Done

File: modules/user/user.graphql.js (hunk 7 above)
Comment (YukiAmeka) [reply modules/user/user.graphql.js:29]: Done

File: typeDefs.js [hunk 12]
```diff
@@ -127,6 +127,19 @@ const typeDefs = gql`
...
+  type UserForComment {
+    email: String!
+    name: String
+    images: ImageSet
+    isAdmin: Boolean
+  }
+
+  type UserRate {
+    user: User!
+    rate: Int
```
Comment (Alan18081) [inline typeDefs.js:139]: It should not be nullable as lastUpdatedDate too

File: typeDefs.js [hunk 13]
```diff
@@ -227,6 +240,17 @@ const typeDefs = gql`
...
   }
+  input UserForCommentInput {
+    email: String!
+    name: String
+    images: ImageSetInput
+    isAdmin: Boolean
+  }
+  input UserRateInput {
+    user: ID!
+    rate: Int
```
Comment (Alan18081) [inline typeDefs.js:251]: It should not be nullable as lastUpdatedDate too

File: typeDefs.js (hunk 12 above)
Comment (YukiAmeka) [reply typeDefs.js:139]: This field is an afterthought for potential use in BI, it's not vital for the work of application. Should default value not be recorded per mongoose model or someone want to modify the collection structure without bringing the website down, an extra non-nullable field will be a pain. Is there a benefit to making it such that I can't see?

File: typeDefs.js (hunk 13 above)
Comment (YukiAmeka) [reply typeDefs.js:251]: This field is an afterthought for potential use in BI, it's not vital for the work of application. Should default value not be recorded per mongoose model or someone want to modify the collection structure without bringing the website down in future, an extra non-nullable field will be a pain. Is there a benefit to making it such that I can't see?

File: typeDefs.js (hunk 12 above)
Comment (YukiAmeka) [reply typeDefs.js]: rate updated, lastUpdatedDate removed

File: typeDefs.js (hunk 13 above)
Comment (YukiAmeka) [reply typeDefs.js]: rate updated, lastUpdatedDate removed

File: typeDefs.js [hunk 14]
```diff
@@ -227,6 +239,16 @@ const typeDefs = gql`
...
     appartment: String
   }
+  input UserForCommentInput {
+    email: String!
+    name: String
+    images: ImageSetInput
+    isAdmin: Boolean
+  }
+  input UserRateInput {
+    user: ID!
```
Comment (Alan18081) [inline typeDefs.js:249]: We don't need an UserId here. It is going to be taken from the jwt token inside app

Comment (Alan18081) [reply typeDefs.js:249]: I think you can get rid of overall UserRateInput itself

Comment (YukiAmeka) [reply typeDefs.js]: Done
